### DISCUSSION TITLE
Resolved: hashes/secp/bitcoin dependency hole

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1985] [General]
A while back we moved `bitcoin_hashes` into `rust-bitcoin/hashes` with the aim of making development and review easier/faster. In doing so we made it very difficult to update `hashes`. In hindsight this was a bad move. Furthermore I postulate that `hashes` actually should never have lived in `rust-bitcoin` because the maintainers of `hashes` should not be the same as `bitcoin` but should be the set that maintains `secp` - `hashes` is crypto code same as `secp`. I am currently a maintainer of `hashes` but I am correctly, not a maintainer of `secp` because I am not a cryptographer, I should not be a maintainer of `hashes`.

### Proposal

- Make `rust-secp256k` crate use a workspace
- Move `hashes` to the secp workspace
- Export `hashes` from `secp256k1`
- In `bitcoin` remove the dependency on `hashes` and use `use secp256k1::hashes`

### Benefits

- Easier release of `bitcoin` and `secp256k1`
- Better maintenance of `hashes` because changes are not lost in the noise of `bitcoin`

**Comment by apoelstra** [comment]
I don't really buy this -- `hashes` is very easy to mainain because it's basically impossible to introduce constant-timeness bugs (for the hashing functions themselves anyway), there's basically no room between "0% correct and 100% correct", and because the code almost never changes. And there's no C bindings, and there's no variable-size structures, and no thread-safety concerns, etc etc etc.

I also don't think we should export `hashes` from `secp`, because the former as a pure-rust not-very-efficient crate that we're kinda lazy about maintaining, and the latter is a set of bindings to an extremely high-quality C library whose maintenance we take quite seriously.

I could maybe be convinced to move `hashes` to the `secp` repo if it makes maintenance easier, though it then does put a bit of an additional burden on the rust-secp maintainers.

**Reply by tcharding** [reply]
Drats, I got super excited about this "solution". Not to worry, cheers.

**Reply by apoelstra** [reply]
On IRC @afilini was proposing that we try to pull the `hashes` sha2 code into rust-secp to replace the existing C hash code, and moving repos would make that possible, so I could still be convinced..

**Comment by tcharding** [comment]
What about if we:

- Move `hashes` to `rust-secp26k1/` workspace
- Add a MAINTAINERS file and some github magic to indicate that it `hashes` requires less review / different reviewers (i.e., just plain software engineers as opposed to crypto geeks)  
- Do not re-export it from secp crate

## Solves
- the release issue (dependency hole thing)
- allows the sha2 code thing

Also, we should really be able to stabalize `hashes` soonish, right? Like you say, we never change it. Perhaps we should audit it and really work out  what needs to be done to push it to 1.0 - I get the feeling we need to get one crate at least over the line else we are never going to get there.

**Reply by apoelstra** [reply]
>Also, we should really be able to stabalize hashes soonish, right?

I dunno. The current API is pretty crappy and the underlying code is undergoing some fairly major revisions (Sanket working on SIMD stuff and Affilini working on code size reduction). I think we're "close" in the sense that it's a pretty small crate but we're not close in the sense of having a clear stabilization plan.

We can certainly audit it and make a plan, but I think @Kixunil will want to be involved.

**Reply by tcharding** [reply]
> We can certainly audit it and make a plan, but I think @Kixunil will want to be involved.

But Kix will make us re-write the whole crate "properly" ( \me starts crying )

**Reply by Kixunil** [reply]
I would definitely want at least raw has/hash newtype split before stabilization. But generally it shouldn't be far. We did most of the work already.

**Comment by tcharding** [comment]
So just to clarify, if we did do this, the move and release process could be:

1. Feature freeze `hashes`
2. Duplicate `hashes` into secp workspace
3. Add patch section to secp workspace to build with local `hashes` (implies update secp code to use newhashes` code)
4. Release `hashes v0.13.0`
5. Update secp to depend on `hashes v0.13.0`
6. Release `secp v0.28.0`
7. Update `bitcoin` to depend on new releases of `secp` and `hashes`
8. Delete `hashes` from `bitcoin`
9. PROFIT

**Comment by tcharding** [comment]
We are almost ready to release the next `hashes` version @apoelstra, I rekon we need to make a decision on the "move to secp repo" thing. Do we move it now or leave it where it is and do the release dance around the dependency hole?

**Reply by apoelstra** [reply]
Let's leave it where it is, but let's remove our internal usage of the `ThirtyTwoByte` trait so that we won't need to dance so much.

**Reply by tcharding** [reply]
Righto, no sweat.

**Reply by Kixunil** [reply]
Oh, that's it! Can we just move the `ThirtyTwoByte` trait into its own pico crate and depend on that one? We could later move it back when we stabilize hashes.

**Reply by apoelstra** [reply]
@Kixunil I'd first like to try dealing with the "hole" by having rust-secp depend on a range of major bitcoin_hashes versions. Because I think this may let us cleanly transition without needing to PR to change dependencies to temporary branches. But I'm not sure because we haven't tried it yet.

**Comment by stevenroose** [comment]
I vote for just leaving hashes as it is right now and not make things complicated. I don't even understand what's the original problem here. Hashes is the simplest crate we have. I could agree it would have been better to just leave it in its own repository, I wasn't active when that happened, but please don't be moving it again and making things even more confusing.

**Reply by apoelstra** [reply]
Yeah, I agree with the sentiment that hashes is the simplest thing we have and if have issues, we should try to fix them elsewhere if possible.

I also believe we may have another solution using ranged dependency versions, though I haven't worked out the details.

But @stevenroose for reference, the issue is that rust-bitcoin depends on rust-secp depends on hashes. But rust-bitcoin has a patch to override hashes when building locally, which does not affect rust-secp since it's in another repo. And this makes it hard to deploy coordinated updates. So it's not hashes that's the issue, but a clash between the repo layout and the dependency tree shape.
